pkg/assembler/backends/inmem/pkg.go

```diff
@@ -421,18 +430,38 @@ func (c *demoClient) IngestPackage(ctx context.Context, input model.PkgInputSpec
 	return c.buildPackageResponse(collectedVersion.id, nil)
 }
 
-func duplicatePkgVer(versions pkgVersionList, input model.PkgInputSpec) (bool, *pkgVersionNode) {
-	for _, v := range versions {
-		if noMatchInput(input.Version, v.version) {
-			continue
-		}
-		if noMatchInput(input.Subpath, v.subpath) {
-			continue
-		}
-		if !reflect.DeepEqual(v.qualifiers, getQualifiersFromInput(input.Qualifiers)) {
-			continue
-		}
-		return true, v
+// hash the canonical representation of a version.
+func hashPkgVersionNode(version *pkgVersionNode) (pkgVersionNodeHash, error) {
+	if version == nil {
+		return "", fmt.Errorf("version is nil")
+	}
+	return hashVersionHelper(version.version, version.subpath, version.qualifiers), nil
+}
+
+// hash the canonical representation of a version
+func hashPkgInputSpecVersion(input model.PkgInputSpec) pkgVersionNodeHash {
+	qualifiers := getQualifiersFromInput(input.Qualifiers)
+	return hashVersionHelper(nilToEmpty(input.Version), nilToEmpty(input.Subpath), qualifiers)
+}
+
+func hashVersionHelper(version string, subpath string, qualifiers map[string]string) pkgVersionNodeHash {
+	// first sort the qualifiers
+	qualifierSlice := []string{}
+	for key, value := range qualifiers {
+		qualifierSlice = append(qualifierSlice, fmt.Sprintf("%s:%s,", key, value))
```

Comment:
In the current implementation, each key-value pair in the qualifiers map is formatted as `%s:%s,` appending a comma after each pair.

This leads to a trailing comma after the last qualifier when all the qualifiers are concatenated. This trailing comma could cause inconsistencies or issues when parsing or comparing the hashed string later.

``` golang
func hashVersionHelper(version string, subpath string, qualifiers map[string]string) pkgVersionNodeHash {
    // First sort the qualifiers.
    qualifierSlice := make([]string, 0, len(qualifiers))
    for key, value := range qualifiers {
        qualifier := fmt.Sprintf("%s:%s", key, value)
        qualifierSlice = append(qualifierSlice, qualifier)
    }
    slices.Sort(qualifierSlice)
    
    // Join the sorted qualifiers with a comma.
    qualifiersStr := strings.Join(qualifierSlice, ",")
    
    // Create the canonical version.
    canonicalVersion := fmt.Sprintf("%s,%s,%s", version, subpath, qualifiersStr)
    
    // Hash the canonical version.
    digest := sha256.Sum256([]byte(canonicalVersion))
    return pkgVersionNodeHash(fmt.Sprintf("%x", digest))
}
```

Reply:
+1 to this, we need to remove the last comma

Reply:
I added this in, thanks!